src/greenlight/runner.clj

```diff
@@ -66,32 +56,54 @@
           (map (fn [v] (v)))))))
 
 
+(defn filter-test-suite
+  [test-suite arguments]
+  (if-let [is-only? (= ":only" (first arguments))]
+    (filter
+      (comp (partial re-matches (re-pattern (second arguments))) name ::test/ns)
+      (find-tests))
+    test-suite))
+
+
+(defn print-test-info
+  "Print out information about the suite of tests."
+  [test-suite options arguments]
+  (let [tests (filter-test-suite test-suite arguments)]
+    (printf "Found %d tests:\n" (count tests))
+    (doseq [test-case tests]
+      (printf "  %s (%d steps)\n"
+              (::test/title test-case)
+              (count (::test/steps test-case))))
+    true))
+
+
 (defn run-tests!
   "Run a collection of tests."
-  [new-system tests options]
-  (when-not (s/valid? ::test/suite tests)
-    (throw (IllegalArgumentException.
-             (str "Invalid test suite configuration: "
-                  (s/explain-str ::test/suite tests)))))
-  (println "Starting test system...")
-  (let [system (component/start (new-system))]
-    (try
-      (binding [test/*report* (partial report/handle-test-event
-                                       {:print-color (not (:no-color options))})]
-        (println "Running" (count tests) "tests...")
-        (let [results (mapv (partial test/run-test! system) tests)]
-          ; TODO: check result spec?
-          (newline)
-          (report-results results options)
-          (when-let [result-path (:output options)]
-            (println "Saving test results to" result-path)
-            ; FIXME: this results in unreadable data because it often includes
-            ; exceptions in the assertion reports.
-            (spit result-path (prn-str results)))
-          ; Successful if every test passed.
-          (every? (comp #{:pass} ::test/outcome) results)))
-      (finally
-        (component/stop system)))))
+  [new-system test-suite options arguments]
```

Comment:
We should at least make `arguments` an optional parameter since we have callers that depend on this arity. `run-all-tests!` below calls the 3 arg version.

What do you think adding this argument into `options` instead?

Reply:
1. I agree with making arguments an optional parameter in run-tests. That I will add in my next commit
2. My first version had it as option (--namespace). But then I realized that clojure devs are likely familiar with :only option from running unit tests. Hence I think Greenlight tests should follow that convention (lein test :only namespace). I will also make it more clear in GoCD test pipelines `java -jar api-tests.jar --output results.edn test :only api-tests.namespace` In this case with clojure.tools.cli, these arrive in arguments vectore as `[test :only api-tests.namespace]`. It felt to me that `:only` argument will be easier to discover and learn that a new option.

Reply:
After thinking about it a bit more, I agree with your #2. Thanks for making `arguments` optional!